consensus/src/block_storage/sync_manager.rs

```diff
@@ -248,9 +249,39 @@ impl BlockStore {
         quorum_certs.extend(
             blocks
                 .iter()
-                .take(num_blocks as usize - 1)
+                .take(blocks.len() - 1)
                 .map(|block| block.quorum_cert().clone()),
         );
+
+        // check if highest_commit_cert is a fork,
+        // if so, we need to fetch it's block as well, to have a proof of commit.
+        if !blocks
+            .iter()
+            .any(|block| block.id() == highest_commit_cert.certified_block().id())
+        {
+            let mut fetched_commit_blocks = retriever
+                .retrieve_block_for_qc(
+                    highest_commit_cert,
+                    1,
+                    highest_commit_cert.commit_info().id(),
+                )
+                .await?;
+
+            assert_eq!(fetched_commit_blocks.len(), 1);
+            let block = fetched_commit_blocks.pop().expect("blocks are empty");
+            assert_eq!(
+                block.id(),
+                highest_commit_cert.certified_block().id(),
+                "Expecting in the retrieval response, for commit certificate fork, first block should be {}, but got {}",
+                highest_commit_cert.certified_block().id(),
+                block.id(),
+            );
+
+            blocks.push(block);
+            quorum_certs.push(highest_commit_cert.clone());
+        }
+
+        assert_eq!(blocks.len(), quorum_certs.len());
```

Comment:
to be extra conservative we can try to construct a LedgerRecoveryData with highest_commit_cert.ledger_info and ensure find_root succeed otherwise give up to avoid data corruption that prevents from restarting

Reply:
this particular check is for the invariant of the code.  we should never have blocks and quorum_certs have different length.

if we fail to fetch either full list, or additional blocks, we will return there. 

I didn't follow what you were saying with LedgerRecoveryData, are you saying that if first fetch succeeds, but second one fails, we want to do something smarter?